Implement the Launch command

jadonn opened the pull request:
This commit ports the launch command from MicroStack to Sunbeam.
 The implementation is different in key ways. It uses the OpenStack
 SDK instead of relying on calls to the CLI. Some of the asset
  checking and logic is different as well.

I tested that the feature still works on snap-openstack as it does in snap-microstack. If you have issues, though, or would like for me to make changes, please let me know! I welcome feedback on how to improve the feature and its implementation.

hemanthnakkina commented on sunbeam/commands/launch.py:
```diff
@@ -0,0 +1,132 @@
+    """
+    console.print("Launching an OpenStack instance ... ")
+    try:
+        conn = openstack.connect(cloud="sunbeam")
```
Cloud name is hardcoded, should be parameterised

jadonn replied:
I made this into a parameter that the user can provide. The default value is sunbeam if no value is provided.

Will that work to resolve this change? :slightly_smiling_face:

hemanthnakkina commented on sunbeam/commands/launch.py:
```diff
@@ -0,0 +1,132 @@
+
+    with console.status("Creating the OpenStack instance ... "):
+        instance_name = petname.Generate()
+        image = conn.compute.find_image("ubuntu-jammy")
```
L#110 - L#113 have hardcoded values, these need to be parameterised.
Having a command to just run an instance with so many hardcoded values doesn't makes sense.

jadonn replied:
Thanks for the feedback! I will add parameters. I hesitated making parameters to start because I felt like that would be re-implementing the `openstack server create` command, which feels redundant to me. However, I can add optional parameters for these values with defaults that match what you would get if you ran the Terraform plan with default options.

As for renaming ubuntu-jammy to ubuntu, I copied that naming convention from the Terraform plan. The Terraform plan's [resource block for the Ubuntu image](https://github.com/javacruft/snap-openstack/blob/feat/tutorial-single-node/cloud/etc/demo-setup/main.tf#LL46C1-L52C2) gives the name "ubuntu-jammy".

javacruft replied:
Indeed but we own both as a project - changes landed today to rename the default image to 'ubuntu'

jadonn replied:
Thanks for the changes! I rebased today and made all the args parameters.

The feature is not working currently, though, and I need input from you all on it. We have a dependency conflict in versions of the OpenStack SDK. I can't build the snap until resolving the conflict, but if I resolve the conflict by upgrading my SDK version the launch command doesn't work anymore.

OpenStack Yoga requires an SDK version of 0.61.*. I have that listed in `requirements.txt`, but the `upper-constraints.txt` file lists an SDK version of 1.0.1, which is for Zed (2023.1 needs 1.0.2). You wouldn't think this would cause problems, but when I run the launch command with SDK version 1.0.1 in `requirements.txt`, the SDK call to add an IP address to the VM instance fails with an error as shown below:
```
ResourceNotFound: 404: Client Error for url: http://10.20.21.10:80/openstack-nova/v2.1/servers/0307aa11-3293-404a-bf03-b3993ae44440/action, The resource could not be found.
```
The server does exist. If I drop back to a snap built with the launch command and 0.61.* the command works again. In addition, if I add `compute_default_microversion: 2.90` to my `clouds.yaml` file that forces the SDK to use the Nova microversion for Yoga instead of the one for Zed.

How do you all want to resolve the dependency conflict? I can increase the version to 1.0.1 and just add the microversion setting to `clouds.yaml` as a workaround for testing, but we will need to increase the SDK version to 1.0.2 when we move to 2023.1.

javacruft replied:
Settingthe default microversion is preferable - can we do that by passing `compute_default_microversion=2.90` as part of the connection parameters?  We can't guarantee that clouds.yaml will have been updated with this information so having something contained in the code is preferable.

javacruft commented:
You'll also need to rebase this branch - thanks @jadonn

javacruft commented on sunbeam/commands/launch.py:
```diff
@@ -0,0 +1,132 @@
+        instance_name = petname.Generate()
+        image = conn.compute.find_image("ubuntu-jammy")
+        flavor = conn.compute.find_flavor("m1.tiny")
+        network = conn.network.find_network("demo-network")
```
If the user running the configure step chooses a username other than demo, this won't work - the network is named {username}-network

jadonn replied:
I made the network option a parameter with a default of `demo-network`. I will make sure to include a note in the documentation on the launch feature that if you change from the defaults in the configure step you must pass in the changed value in the launch step.

Will that be good enough or should I take a different approach?

hemanthnakkina replied:
Please speak with @javacruft, I personally feel the approach should be f"{tf_output["OS_USERNAME"]["value"]}-network"

javacruft commented:
@jadonn with regards to the openstacksdk version - is this just a matter of setting the required microversions for yoga vs antelope? if so I'd rather stick with the openstacksdk from upper-constraints and set the needed microversion for now.

javacruft commented:
OK I think we need to hit reset with the approach to the launch command - this is getting way to complicated.

We need an easy way for someone who has run the install with --accept-defaults (or at least created a demonstration user and project using the guided configuration process) to be able to quick and easily launch an instance.

We can't guarantee that they have run the cloud-config command to update/create the clouds.yaml and TBH we don't need to - all of the demonstration user access details are in the output of the terraform plan execution.

If the demonstration user and project has not be created,  launch should fail its pre-flight check.

I think this should work:

```
sunbeam launch ubuntu
```

which means - launch me a new instance of the ubuntu image and give me access - sunbeam can figure out the details.

maybe optionally they could provide a name

```
sunbeam launch --name foobar ubuntu
```

same as above but call it foobar in nova.

Anyone more complicated - run cloud-config and use the bundled openstack client.

jadonn commented:
I'm fine with reducing the launch command to as small a bit of functionality as we can manage and with encouraging users to use the CLI for anything more complex than a simple demonstration use case.

What you have described is essentially what my first implementation/submission aimed to accomplish. It was even simpler than what you have described; the command was originally `sunbeam launch` with no args. It also failed if there was no clouds.yaml or authentication information, and it prompted the user to run the configure step.

Yeah, the values for the instance creation were hard-coded, but as I understood it, we only wanted it to work if the user followed the demonstration journey with `--accept-defaults`. So I hard-coded the arguments for the instance attributes to match what the Terraform plan had. It may be worthwhile to pull the values for Terraform and for the launch command to a shared source to ensure they stay synchronized. We're already hard-coding these things in the Terraform plan at this time.

I can revert the changes I have made to make parameters out of everything and work from what I originally submitted to make it like what you have described.

javacruft commented:
Where parameters can be identified as fixed, I'm OK with coded literals as both pieces sit in the same repo.

The terraform plan for the demonstration user is parameterised and and has a defined set of outputs - @hemanthnakkina uses these outputs in the [cloud-config command](https://github.com/openstack-snaps/snap-openstack/commit/106d25ea40a3b592cd64e8079935f78bc8b4c667) we've just landed to pull out the details for the user; you can follow the same pattern to get access details for the cloud.

Using this approach allows us to support both the --accept-defaults route and the guided configuration route to a deployed cloud.

jadonn commented:
OK. I reimplemented the `launch` command. It is narrowly constrained to fit what we have discussed for what the `launch` command should do. Let me know what improvements could be made. I am happy to make adjustments and changes.

There is one question I could use some input on. In order to facilitate generating an SSH key for logging into the instance, I added another personal-files plug and app called `dot-ssh` for having write access to `~/.ssh`. I implemented the `launch` command to write a private key to `~/.ssh` if the argument for `key` does not exist in OpenStack already. Is adding this plug all right with you all?

hemanthnakkina commented on sunbeam/commands/launch.py:
```diff
@@ -0,0 +1,159 @@
+    "-k",
+    "--key",
+    default="sunbeam",
+    help="The path to the SSH key to use for the instance"
```
Can you change help to reflect this option is for key name (instead of path to SSH key) and about creation of key pair with private key written to specific directory.

javacruft replied:
:+1: to @hemanthnakkina comment above - this is the name of the key to create in OpenStack

jadonn replied:
I updated this to reflect the name of the key in OpenStack to use and that a key is created if it does not exist.

hemanthnakkina commented:
> OK. I reimplemented the `launch` command. It is narrowly constrained to fit what we have discussed for what the `launch` command should do. Let me know what improvements could be made. I am happy to make adjustments and changes.
> 
> There is one question I could use some input on. In order to facilitate generating an SSH key for logging into the instance, I added another personal-files plug and app called `dot-ssh` for having write access to `~/.ssh`. I implemented the `launch` command to write a private key to `~/.ssh` if the argument for `key` does not exist in OpenStack already. Is adding this plug all right with you all?

Can we reuse ~/.config/openstack directory instead of getting write access to ~/.ssh

gboutry commented on sunbeam/commands/launch.py:
```diff
@@ -0,0 +1,159 @@
+            " Have you run the configure command?",
+            " Do you have a clouds.yaml file?",
+        )
+        return
```
This should return an exit code different from 0, since it's a failure

javacruft replied:
This exception block is dealing with to much - we should never get to the point of trying to launch an instances if the configure process has not been executed, and that during that process the user has not setup a demonstration user, project and network.

Using:

```python
raise click.ClickException(f"Unable to connect to OpenStack: {e.message}")
```

covers @gboutry feedback (exit 1 will occur) and provides the user with some sort of hint about what actually went wrong.

gboutry commented:
> > OK. I reimplemented the `launch` command. It is narrowly constrained to fit what we have discussed for what the `launch` command should do. Let me know what improvements could be made. I am happy to make adjustments and changes.
> > There is one question I could use some input on. In order to facilitate generating an SSH key for logging into the instance, I added another personal-files plug and app called `dot-ssh` for having write access to `~/.ssh`. I implemented the `launch` command to write a private key to `~/.ssh` if the argument for `key` does not exist in OpenStack already. Is adding this plug all right with you all?
> 
> Can we reuse ~/.config/openstack directory instead of getting write access to ~/.ssh

Moreover, we already tried getting access to dot-ssh, but it was declined as it would be too dangerous.

javacruft commented on sunbeam/commands/launch.py:
```diff
@@ -0,0 +1,159 @@
+    with console.status("Allocating IP address to instance ... "):
+        external_network = conn.network.find_network("external-network")
+        ip_ = conn.network.create_ip(floating_network_id=external_network.id)
+        conn.compute.add_floating_ip_to_server(server_id, ip_.floating_ip_address)
```
Please can we wrap all of the above in an exception handler for any SDK errors thrown in the same way you did for the connect call.

jadonn replied:
Done! I wrapped that in an exception handler, and I also added one to the VM instance creation to catch exceptions.

javacruft commented on sunbeam/commands/launch.py:
```diff
@@ -0,0 +1,159 @@
+            console.print(f"No {key} key found in OpenStack. Creating SSH key at {key_path}")
+            key_id = conn.compute.create_keypair(name=key)
+            with open(key_path, "w", encoding="utf-8") as key_file:
+                key_file.write(key_id.private_key)
```
Rather than using console.print in the above code block, you can update the status object that is returned by the with call and simply use `.update` - for example:

```python
status.update("Checking for SSH public key in OpenStack ...")
```

this will display it with the spinner rather than it being persisted on screen.

jadonn replied:
Thank you for the advice! I made those changes as you suggested.

javacruft commented:
One final comment - we should do a check as early as possible to determine a) whether the cloud is even bootstrapped and b) whether the demonstration user setup has been completed as part of configure - if not then a polite message to the user to this effect is in order.

jadonn commented on sunbeam/commands/launch.py:
```diff
@@ -0,0 +1,172 @@
+        id_ = openstack_conn.compute.create_keypair(name="sunbeam")
+        with open(key_path, "w", encoding="utf-8") as file_:
+            file_.write(id_.private_key)
+            check("chmod", "600", key_path)
```
I'm resolving this conversation since this is outdated now.

jadonn commented on requirements.txt:
```diff
@@ -25,3 +25,7 @@ pyyaml>=6.0
 #juju>=3.2,<3.3
+
+# Used in the launch command to launch an instance
+openstacksdk==1.0.1
```
OK! I have removed this constraint.

jadonn commented on snap/snapcraft.yaml:
```diff
@@ -249,6 +250,11 @@ plugs:
+    interface: personal-files
+    write:
+    - $HOME/.ssh
+
```
OK! That is good to know. Thank you! I have removed this.

jadonn commented on sunbeam/commands/launch.py:
```diff
@@ -0,0 +1,159 @@
+def launch(
+    image_name: str = "ubuntu",
+    key: str = "sunbeam",
+    name: str = ""
```
Thanks! I am not sure what the current practice in this area is now, so I appreciate the guidance. I'm still new to typings in Python as well. It's neat seeing this come into Python from other languages.

jadonn commented on sunbeam/commands/launch.py:
```diff
@@ -0,0 +1,159 @@
+
+        action_result = run_sync(jhelper.run_action(unit, OPENSTACK_MODEL, action_cmd))
+        if action_result.get("return-code", 0) > 1:
+            _message = "Unable to retrieve openrc from Keystone service"
```
I switched to using `retrieve_admin_credentials`. While I understand the value of not repeating yourself, this does introduce a dependency on the `configure` command into the `launch` command. I wanted to point out that this could be problematic in the future if `retrieve_admin_credentials` changes in some way that negatively impacts the `launch` command. I'm also not sure of what reason, other than not repeating the implementation, there is for introducing such a dependency. Sure, the configure command must run before the launch command, but I feel the commands are otherwise independent of one another.

jadonn commented on sunbeam/commands/launch.py:
```diff
@@ -0,0 +1,159 @@
+        return
+
+    with console.status("Checking for SSH key pair ... "):
+        key_path = f"{home}/.ssh/{key}"
```
I removed this and changed the logic to save the key in ~/.config/openstack/ like Hemanth suggested

gboutry commented on sunbeam-python/sunbeam/commands/launch.py:
```diff
@@ -0,0 +1,151 @@
+)
+@click.option("-n", "--name", help="The name for the instance.")
+def launch(image_name: str, key: str, name: Optional[str] = None) -> None:
+    """Launch an OpenStack instance"""
```
As it is, the launch command can only create an instance on the demo project. I believe this should be at least specified in the command's help

hemanthnakkina reviewed (approved):
LGTM, the command help can be fixed in subsequent PR.
